### PR TITLE
Add Y2038 design

### PR DESCRIPTION
**Pull request by sbomer**

### CODE REVIEW

File: docs/design/y2038.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
+# Y2038 support in .NET 9 linux arm32
+
+In .NET 9, we addressed the Y2038 problem for linux arm32. mattheweshleman [noticed](https://github.com/dotnet/runtime/issues/96460) the issue early in the release cycle, and we also [hit](https://github.com/dotnet/runtime/issues/101444) this when our linux arm32 builds started hitting openssl-related errors on Ubuntu 24.04, the first Ubuntu version to enable 64-bit `time_t` by default for linux arm32. This document describes our approach to Y2038 support.
```
Comment (jkotas) [inline docs/design/y2038.md:3]:
```suggestion
In .NET 9, we addressed the Y2038 problem for linux arm32. @mattheweshleman [noticed](https://github.com/dotnet/runtime/issues/96460) the issue early in the release cycle, and we also [hit](https://github.com/dotnet/runtime/issues/101444) this when our linux arm32 builds started hitting openssl-related errors on Ubuntu 24.04, the first Ubuntu version to enable 64-bit `time_t` by default for linux arm32. This document describes our approach to Y2038 support.
```

File: docs/design/y2038.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+[Ubuntu](https://discourse.ubuntu.com/t/ubuntu-24-04-lts-noble-numbat-release-notes/39890#year-2038-support-for-the-armhf-architecture-5) made this change in 24.04, and [Debian](https://wiki.debian.org/ReleaseGoals/64bit-time) 13 is planned to be the first Debian release with 64-bit time.
+
+[musl](https://musl.libc.org/time64.html) version 1.2.0 got 64-bit time_t support, which was adopted in [Alpine](https://wiki.alpinelinux.org/wiki/Release_Notes_for_Alpine_3.13.0) 3.13.
+
+### .NET builds
+
+.NET official builds are produced by building the product on a dedicated set of build images. The images run Azure Linux 3.0 with a modern cross-compilation toolchain, and also include a root filesystem of an old linux distribution that provides an old version of glibc (or musl libc), which determines the libc compatibility of our builds.
```
Comment (jkotas) [inline docs/design/y2038.md:27]:
```suggestion
.NET official builds are produced by building the product on a dedicated set of build images. The images run Azure Linux 3.0 with an up-to-date cross-compilation toolchain, and also include a root filesystem of an old linux distribution that provides an old version of glibc (or musl libc), which determines the libc compatibility of our builds.
```
I would not describe the toolchain as modern.

File: docs/design/y2038.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
+# Y2038 support in .NET 9 linux arm32
```
Comment (jkotas) [inline docs/design/y2038.md:1]:
```suggestion
# Y2038 support in .NET 9 Linux arm32
```
Linux is a name of a thing. Start with upper case? (multiple places)

File: docs/design/y2038.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+In .NET 9, we addressed the Y2038 problem for linux arm32. mattheweshleman [noticed](https://github.com/dotnet/runtime/issues/96460) the issue early in the release cycle, and we also [hit](https://github.com/dotnet/runtime/issues/101444) this when our linux arm32 builds started hitting openssl-related errors on Ubuntu 24.04, the first Ubuntu version to enable 64-bit `time_t` by default for linux arm32. This document describes our approach to Y2038 support.
+
+### Summary
+
+Our default posture when facing ABI breaks in the Linux ecosystem is to first align with the latest distributions, and then work backwards to determine a reasonable compatibility story for older distributions.
```
Comment (jkotas) [inline docs/design/y2038.md:7]:
```suggestion
Our default posture for latest .NET is to align with the latest Linux distributions first, and then work backwards to determine a reasonable compatibility story for older distributions.
```
This is not limited to ABI breaks

File: docs/design/y2038.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+.NET 8 already [supports](https://github.com/dotnet/runtime/pull/51099) 64-bit time on musl linux, and on architectures other than arm. This just leaves the glibc-based arm32 linux build.
+
+#### glibc
+
+Because our build already [defined](https://github.com/dotnet/runtime/pull/100461) `_TIME_BITS=64`, the main change we had to make was to [update](https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/1037) the Ubuntu version of our rootfs to one that had a glibc with `_TIME_BITS` support. By targeting Ubuntu 22.04, .NET would be built with 64-bit `time_t`, allowing `DateTime.UtcNow` to [work](https://github.com/dotnet/runtime/issues/96460) correctly past Y2038.
+
+With this change, glibc-based .NET 9 (on linux arm) is only [supported](https://github.com/dotnet/core/blob/main/release-notes/9.0/supported-os.md) on distributions with at least glibc 2.35 (from Ubuntu 22.04).
+
+#### openssl
```
Comment (jkotas) [inline docs/design/y2038.md:41]:
```suggestion
#### OpenSSL
```
?

**Review by jkotas — APPROVED** [review]
Thanks!

File: docs/design/y2038.md [hunk 6]
```diff
@@ -0,0 +1,91 @@
+# Y2038 support in .NET 9 Linux arm32
+
+In .NET 9, we addressed the Y2038 problem for Linux arm32. @mattheweshleman [noticed](https://github.com/dotnet/runtime/issues/96460) the issue early in the release cycle, and we also [hit](https://github.com/dotnet/runtime/issues/101444) this when our Linux arm32 builds started hitting OpenSSL-related errors on Ubuntu 24.04, the first Ubuntu version to enable 64-bit `time_t` by default for Linux arm32. This document describes our approach to Y2038 support.
+
+### Summary
+
+Our default posture for latest .NET to align with the latest Linux distribution first, and then work backwards to determine a reasonable compatibility story for older distributions, so .NET 9 Linux arm32 requires glibc 2.35 and uses 64-bit `time_t` by default. This works on systems with 32-bit `time_t` as long as we don't need to represent times larger than 32 bits. We detect whether OpenSSL uses 32-bit `time_t`, and fail predictably if the time being represented on the .NET side does not fit in 32 bits.
```
Comment (jkotas) [inline docs/design/y2038.md:7]:
```suggestion
Our default posture for latest .NET is to align with the latest Linux distribution first, and then work backwards to determine a reasonable compatibility story for older distributions, so .NET 9 Linux arm32 requires glibc 2.35 and uses 64-bit `time_t` by default. This works on systems with 32-bit `time_t` as long as we don't need to represent times larger than 32 bits. We detect whether OpenSSL uses 32-bit `time_t`, and fail predictably if the time being represented on the .NET side does not fit in 32 bits.
```

File: docs/design/y2038.md [hunk 7]
```diff
@@ -0,0 +1,91 @@
+# Y2038 support in .NET 9 Linux arm32
+
+In .NET 9, we addressed the Y2038 problem for Linux arm32. @mattheweshleman [noticed](https://github.com/dotnet/runtime/issues/96460) the issue early in the release cycle, and we also [hit](https://github.com/dotnet/runtime/issues/101444) this when our Linux arm32 builds started hitting OpenSSL-related errors on Ubuntu 24.04, the first Ubuntu version to enable 64-bit `time_t` by default for Linux arm32. This document describes our approach to Y2038 support.
```
Comment (richlander) [inline docs/design/y2038.md:3]:
```suggestion
In .NET 9, we addressed the Y2038 problem for Linux arm32. @mattheweshleman [noticed](https://github.com/dotnet/runtime/issues/96460) the issue early in the release cycle, and we also [hit](https://github.com/dotnet/runtime/issues/101444) this when our Linux arm32 builds started hitting OpenSSL-related errors on Ubuntu 24.04, the first Ubuntu version to [enable 64-bit `time_t` by default for Linux arm32](https://discourse.ubuntu.com/t/ubuntu-24-04-lts-noble-numbat-release-notes/39890#year-2038-support-for-the-armhf-architecture-5). This document describes our approach to Y2038 support.
```

File: docs/design/y2038.md [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+### Summary
+
+Our default posture for latest .NET is to align with the latest Linux distribution first, and then work backwards to determine a reasonable compatibility story for older distributions, so .NET 9 Linux arm32 requires glibc 2.35 and uses 64-bit `time_t` by default. This works on systems with 32-bit `time_t` as long as we don't need to represent times larger than 32 bits. We detect whether OpenSSL uses 32-bit `time_t`, and fail predictably if the time being represented on the .NET side does not fit in 32 bits.
```
Comment (richlander) [inline docs/design/y2038.md:7]:
```suggestion
Our default posture for new .NET versions is to align with the latest Linux distribution first, and then work backwards to determine a reasonable compatibility story for older distributions. For .NET 9 Linux arm32 build require glibc 2.35 and use 64-bit `time_t` by default. We've made this change to align with the Ubuntu 24.04. This change works on systems with 32-bit `time_t` but cannot represent times larger than 32 bits. We detect whether OpenSSL uses 32-bit `time_t` and fail predictably if the time being represented on the .NET side does not fit in 32 bits.
```

File: docs/design/y2038.md [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+[Ubuntu](https://discourse.ubuntu.com/t/ubuntu-24-04-lts-noble-numbat-release-notes/39890#year-2038-support-for-the-armhf-architecture-5) made this change in 24.04, and [Debian](https://wiki.debian.org/ReleaseGoals/64bit-time) 13 is planned to be the first Debian release with 64-bit time.
+
+[musl](https://musl.libc.org/time64.html) version 1.2.0 got 64-bit time_t support, which was adopted in [Alpine](https://wiki.alpinelinux.org/wiki/Release_Notes_for_Alpine_3.13.0) 3.13.
+
+### .NET builds
+
+.NET official builds are produced by building the product on a dedicated set of build images. The images run Azure Linux 3.0 with an up-to-date cross-compilation toolchain, and also include a root filesystem of an old Linux distribution that provides an old version of glibc (or musl libc), which determines the libc compatibility of our builds.
```
Comment (richlander) [inline docs/design/y2038.md:25]:
```suggestion
.NET official builds are produced by building the product on a [dedicated set of build images](https://github.com/dotnet/runtime/blob/main/docs/workflow/building/coreclr/linux-instructions.md). The images run Azure Linux 3.0 with an up-to-date cross-compilation toolchain, and also include a root filesystem of an old Linux distribution that provides an old version of glibc (or musl libc), which determines the libc compatibility of our builds.
```

File: docs/design/y2038.md [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+In .NET 8, we [support](https://github.com/dotnet/core/blob/main/release-notes/8.0/supported-os.md#libc-compatibility) glibc 2.23 andc musl 1.2.2, by building the product with a root filesystem from Ubuntu 16.04 and Alpine 3.13, respectively.
```
Comment (richlander) [inline docs/design/y2038.md:27]:
```suggestion
In .NET 8, we [support](https://github.com/dotnet/core/blob/main/release-notes/8.0/supported-os.md#libc-compatibility) glibc 2.23 and musl 1.2.2, by building the product with a root filesystem from Ubuntu 16.04 and Alpine 3.13, respectively.
```

File: docs/design/y2038.md [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+
+### .NET Y2038 support
+
+.NET 8 already [supports](https://github.com/dotnet/runtime/pull/51099) 64-bit time on musl Linux, and on architectures other than arm. This just leaves the glibc-based arm32 Linux build.
```
Comment (richlander) [inline docs/design/y2038.md:31]:
```suggestion
.NET 8 already [supports](https://github.com/dotnet/runtime/pull/51099) 64-bit time on musl Linux, and on architectures other than 32-bit arm. This just leaves the glibc-based arm32 Linux build.
```

**Review by richlander — APPROVED** [review]

**Comment by richlander** [comment]
Thanks!

**Pull request merged**